src/core.c/Any-iterable-methods.pm6

```diff
@@ -2357,6 +2357,62 @@ Consider using a block if any of these are necessary for your mapping code."
         Seq.new(Rakudo::Iterator.Rotor(self.iterator,@cycle,$partial))
     }
 
+    proto method infer(|) {*}
+    multi method infer(List:D:) {
+        my $iterator := self.iterator;
+        nqp::if(
+          nqp::eqaddr((my $pulled := $iterator.pull-one),IterationEnd),
+          Nil,                                            # nothing to check
+          nqp::stmts(
+            (my $type := $pulled.WHAT),                   # initial type
+            nqp::until(
+              nqp::eqaddr(($pulled := $iterator.pull-one),IterationEnd)
+                || nqp::not_i(nqp::eqaddr($pulled.WHAT,$type)),
+              nqp::null
+            ),
+            nqp::if(
+              nqp::eqaddr($pulled,IterationEnd),
+              $type,                                      # all the same
+              self!slow-infer($iterator, $type, $pulled)  # find out what
+            )
+          )
+        )
+    }
+
+    method !slow-infer($iterator, Mu $type is copy, Mu $pulled is copy) {
+        # set up types to check
+        my $mro   := nqp::clone(nqp::getattr($type.^mro,List,'$!reified'));
+        my $roles := nqp::clone(nqp::getattr($type.^roles,List,'$!reified'));
```

Comment:
Wouldn't `.^mro(:roles)` simplify the code?

Reply:
That doesn't seem to work?
````
$ raku -e 'dd Int.^mro(:roles)'
Cannot find method 'find_method' on object of type Perl6::Metamodel::ConcreteRoleHOW
````

Reply:
If that would work, and the roles be in the right order, yes, that would simplify the code quite a bit  :-)

Reply:
It works, but - concrete role... Hm, they're functionally limited, that's true. Considering the most popular pattern of `mro` use, I should probably rework it to return a parametric group.

Reply:
> If that would work, and the roles be in the right order, yes, that would simplify the code quite a bit :-)

The order is right, `Int.^mro(:roles).map(*.^name)` gives `(Int Real Numeric Cool Any Mu)`. But there is a little problem. This parameter's primary purpose is to support 6.e semantics of submethods. It has to provide a list of typeobjects suitable for looking up submethods in them. Concretizations are the best candidates for this.

Perhaps it'd be better to have two named variants for each purpose: `:roles` and `:concretizations`. I need to think over it.

Reply:
@lizmat  #4635 implements more relevant `.^mro(:roles)` semantics.

Reply:
@lizmat #4635 is merged, you could try simplifying your algorithm.